bridges/snowbridge/primitives/core/src/reward.rs

```diff
@@ -1,40 +1,406 @@
+		match e {
+			RewardPaymentError::XcmSendFailure => DispatchError::Other("xcm send failure"),
+			RewardPaymentError::ChargeFeesFailure => DispatchError::Other("charge fees error"),
+			RewardPaymentError::XcmDeliveryFailure => DispatchError::Other("xcm delivery failure"),
+		}
+	}
 }
 
-impl<AccountId> PaymentProcedure<AccountId> for () {
-	type Error = &'static str;
+pub struct NoOpReward;
+/// Reward payment procedure that sends a XCM to AssetHub to mint the reward (foreign asset)
+/// into the provided beneficiary account.
+pub struct PayAccountOnLocation<
+	Relayer,
+	RewardBalance,
+	NoOpReward,
+	EthereumNetwork,
+	AssetHubLocation,
+	AssetHubXCMFee,
+	InboundQueueLocation,
+	XcmSender,
+	XcmExecutor,
+	Call,
+>(
+	PhantomData<(
+		Relayer,
+		RewardBalance,
+		NoOpReward,
+		EthereumNetwork,
+		AssetHubLocation,
+		AssetHubXCMFee,
+		InboundQueueLocation,
+		XcmSender,
+		XcmExecutor,
+		Call,
+	)>,
+);
+
+impl<
+		Relayer,
+		RewardBalance,
+		NoOpReward,
+		EthereumNetwork,
+		AssetHubLocation,
+		AssetHubXCMFee,
+		InboundQueueLocation,
+		XcmSender,
+		XcmExecutor,
+		Call,
+	> PaymentProcedure<Relayer, NoOpReward, RewardBalance>
+	for PayAccountOnLocation<
+		Relayer,
+		RewardBalance,
+		NoOpReward,
+		EthereumNetwork,
+		AssetHubLocation,
+		AssetHubXCMFee,
+		InboundQueueLocation,
+		XcmSender,
+		XcmExecutor,
+		Call,
+	>
+where
+	Relayer: Clone
+		+ Debug
+		+ Decode
+		+ Encode
+		+ Eq
+		+ TypeInfo
+		+ Into<sp_runtime::AccountId32>
+		+ Into<Location>,
+	EthereumNetwork: Get<NetworkId>,
+	InboundQueueLocation: Get<InteriorLocation>,
+	AssetHubLocation: Get<Location>,
+	AssetHubXCMFee: Get<u128>,
+	XcmSender: SendXcm,
+	RewardBalance: Into<u128> + Clone,
+	XcmExecutor: ExecuteXcm<Call>,
+	Call: Decode + GetDispatchInfo,
+{
+	type Error = DispatchError;
+	type Beneficiary = Location;
 
 	fn pay_reward(
-		_: AccountId,
-		_: Asset,
+		relayer: &Relayer,
+		_reward_kind: NoOpReward,
+		reward: RewardBalance,
+		beneficiary: Self::Beneficiary,
 	) -> Result<(), Self::Error> {
+		let ethereum_location = Location::new(2, [GlobalConsensus(EthereumNetwork::get())]);
+
+		let total_amount: u128 = AssetHubXCMFee::get().saturating_add(reward.clone().into());
+		let total_assets: Asset = (ethereum_location.clone(), total_amount).into();
+		let fee_asset: Asset = (ethereum_location, AssetHubXCMFee::get()).into();
+
+		let xcm: Xcm<()> = alloc::vec![
+			DescendOrigin(InboundQueueLocation::get().into()),
+			UniversalOrigin(GlobalConsensus(EthereumNetwork::get())),
+			ReserveAssetDeposited(total_assets.into()),
+			PayFees { asset: fee_asset },
+			DepositAsset { assets: AllCounted(1).into(), beneficiary },
+			RefundSurplus,
+		]
+		.into();
+
+		let (ticket, fee) =
+			validate_send::<XcmSender>(AssetHubLocation::get(), xcm).map_err(|_| XcmSendFailure)?;
+		XcmExecutor::charge_fees(relayer.clone(), fee.clone()).map_err(|_| ChargeFeesFailure)?;
+		XcmSender::deliver(ticket).map_err(|_| XcmDeliveryFailure)?;
```

Comment:
`XcmSendFailure` and `XcmDeliveryFailure` mean the same thing, just choose one of them.

```suggestion
		XcmSender::deliver(ticket).map_err(|_| XcmSendFailure)?;
```

Reply:
Adressed in https://github.com/vgeddes/polkadot-sdk/pull/12/commits/53da53cffa35dad43ffe333121b60e6120e4490a.